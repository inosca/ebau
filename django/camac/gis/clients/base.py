from typing import Any

from caluma.caluma_form.models import Question
from django.http import QueryDict

from camac.gis.utils import MergeStrategy, concat_values, merge_table


class GISBaseClient:
    merge_strategy: MergeStrategy = MergeStrategy.MERGE_FIRST

    def __init__(self, params: QueryDict):
        self.params = params

    @staticmethod
    def get_required_params(data_source):
        return []

    @staticmethod
    def get_hidden_questions(config):
        """Return hidden question slugs based on client config.

        By default, this feature is not enabled. Override this method in
        your client to use it, see e.g. clients/param.py.
        """
        return []

    def process_data_source(self, config: dict, intermediate_data) -> dict:
        raise NotImplementedError()

    def set_question_value(self, data: dict, question: str, value: Any) -> dict:
        if "." in question:
            table_question, row_question = question.split(".")
            row_data = {row_question: value}

            if table_question in data:
                data[table_question] = merge_table(
                    data[table_question], row_data, self.merge_strategy
                )
            else:
                data[table_question] = [row_data]

            return

        question_obj = Question.objects.filter(slug=question).first()
        if not question_obj:
            raise RuntimeError(f"Unknown question {question} in gis config")
        if type(value) == list and question_obj.type != Question.TYPE_MULTIPLE_CHOICE:
            value = ", ".join(sorted(set(value)))

        if question in data:
            value = concat_values(data[question], value)

        data[question] = value
