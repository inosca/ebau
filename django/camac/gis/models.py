from django.db import models
from django.utils.translation import gettext_lazy as _


class GISConfig(models.Model):
    CLIENT_SOGIS = "camac.gis.clients.sogis.SoGisClient"
    CLIENT_PARAM = "camac.gis.clients.param.ParamClient"

    CLIENT_CHOICES = [
        (CLIENT_SOGIS, _("GIS Canton Solothurn")),
        (CLIENT_PARAM, _("Parameter")),
    ]

    slug = models.SlugField(primary_key=True)
    client = models.CharField(max_length=255, choices=CLIENT_CHOICES)
    config = models.JSONField(default=dict)
    disabled = models.BooleanField(default=False)
