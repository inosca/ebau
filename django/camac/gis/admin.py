from adminsortable2.admin import SortableAdminMixin
from django.contrib.admin import ModelAdmin, register
from django.db.models import JSONField
from django_json_widget.widgets import JSONEditorWidget

from camac.admin import EbauAdminMixin
from camac.gis.models import GISDataSource


@register(GISDataSource)
class GISDataSourceAdmin(EbauAdminMixin, SortableAdminMixin, ModelAdmin):
    list_display = ["description", "client", "disabled"]
    list_per_page = 50
    search_fields = ["description", "client"]
    formfield_overrides = {
        JSONField: {"widget": JSONEditorWidget},
    }
