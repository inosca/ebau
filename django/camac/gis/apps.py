from django.apps import AppConfig


class GisConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "camac.gis"
