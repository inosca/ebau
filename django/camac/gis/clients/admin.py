import requests
from django.conf import settings
from django.utils.translation import gettext as _

from camac.gis.clients.base import GISBaseClient
from camac.gis.utils import cast, concat_values, get_bbox, to_query
from camac.utils import build_url


class AdminGisClient(GISBaseClient):
    required_params = ["x", "y"]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.session: requests.Session = requests.Session()

    def process_data_source(self, config: dict) -> dict:
        """Process admin GIS data source.

        Example config:
        {
            "layers": ["ch.some_layername"],
            "buffer": 100,
            "attributes": [
                { "attributeName": "property_name_1", "question": "pathto.myquestion1" },
                { "attributeName": "property_name_2", "question": "pathto.myquestion2", "cast": "integer" },
                { "attributeName": "property_name_2", "question": "pathto.myquestion2", "template": "Description: {value}" }
            ]
        }
        """
        base_url = build_url(
            settings.ADMIN_GIS_BASE_URL,
            "/rest/services/api/MapServer/identify",
        )

        query = to_query(
            {
                "returnGeometry": False,
                "sr": 2056,
                "mapExtent": "0,0,0,0",
                "tolerance": 0,
                "imageDisplay": "0,0,0",
                "layers": "all:" + ",".join(config["layers"]),
                "geometryType": "esriGeometryEnvelope",
                "geometry": get_bbox(
                    self.params["x"],
                    self.params["y"],
                    config.get("buffer", 0),
                ),
            }
        )

        response = self.session.get(
            f"{base_url}?{query}",
            verify=settings.ADMIN_GIS_VERIFY_SSL,
        )

        try:
            response.raise_for_status()
        except requests.HTTPError:
            raise RuntimeError(
                _("Error %(code)s while fetching data from the geo.admin.ch API")
                % {"code": response.status_code}
            )

        try:
            results = response.json()["results"]
        except KeyError:
            return {}

        data = {}

        for attribute_config in config["attributes"]:
            value = None

            for result in results:
                attributes = result["attributes"]
                label = result["layerName"]

                value = concat_values(
                    value, self.get_value(attributes, label, attribute_config)
                )

            if attribute_config["question"] in data:
                value = concat_values(data[attribute_config["question"]], value)

            data[attribute_config["question"]] = value

        return data

    def get_value(self, attributes, label, attribute_config):
        cast_to = attribute_config.get("cast")
        raw_value = cast(
            attributes.get(attribute_config["attributeName"], None), cast_to
        )

        if not raw_value:
            return None

        return (
            attribute_config.get("template", "{value}").format(
                label=label,
                value=raw_value,
            )
            if not cast_to or cast_to == "string"
            else raw_value
        )
