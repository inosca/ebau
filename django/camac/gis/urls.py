from django.urls import re_path

from camac.gis import views

urlpatterns = [
    re_path(
        r"^data/(?:(?P<task_id>[\w-]+)/)?$",
        views.GISDataView.as_view(),
        name="gis-data",
    ),
    re_path(
        r"^apply$",
        views.GISApplyView.as_view(),
        name="gis-apply",
    ),
]
