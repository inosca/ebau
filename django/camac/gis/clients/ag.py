import json
from decimal import Decimal

import requests
from django.conf import settings
from django.utils.translation import gettext as _

from camac.gis.clients.base import GISBaseClient
from camac.utils import build_url


class AgGisClient(GISBaseClient):
    required_params = ["query"]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.session: requests.Session = requests.Session()

    def process_data_source(self, config, _intermediate_data) -> dict:
        """Process AG GIS config."""

        query = json.loads(self.params.get("query"))
        markers = query["markers"]

        point_response_json = self.get_agis_identify_response(markers, 1)
        area_response_json = self.get_agis_identify_response(markers, 20)

        agis_info_from_point = point_response_json.get("results", [])
        agis_info_from_area = area_response_json.get("results", [])
        result = {}

        for layer in config:
            mapper = layer.get("mapper", "properties")
            agis_data = (
                agis_info_from_area
                if layer.get("in_20m_radius", False)
                else agis_info_from_point
            )
            if mapper == "exists":
                question = layer["question"]
                result_answer = self.get_result_for_exists_mapper(
                    result, agis_data, layer
                )
                if result_answer:
                    result[question] = result_answer

            else:
                layers = self.get_all_layers_by_identifier(
                    agis_data, layer["identifier"]
                )
                for layer_info in layers:
                    for prop in layer["properties"]:
                        question = prop["question"]
                        result[question] = self.get_result_for_properties_mapper(
                            result, prop, layer_info
                        )

        return {
            **result,
            "parzelle": self.get_plot_data(agis_info_from_point),
            "gebaeude": self.get_building_data(agis_info_from_point),
        }

    def get_result_for_properties_mapper(self, current_result, property, layer_info):
        property_name = property["propertyName"]
        question = property["question"]
        template = property.get("template", None)
        prop_val = layer_info["attributes"].get(property_name, "")
        new_value = template.replace("{{property}}", prop_val) if template else prop_val
        if new_value in current_result.get(question, ""):
            return current_result.get(question, "")
        return self.get_concatenated_answer(current_result, question, new_value)

    def get_result_for_exists_mapper(self, current_result, agis_data, layer):
        choiceType = layer["type"]
        question = layer["question"]
        answer = layer["answer"]
        answer_else = layer.get("answer_else", None)
        result_answer = None
        identifiers = layer.get("identifiers", [layer.get("identifier")])
        exists = self.layer_exists(agis_data, identifiers)

        if exists:
            if choiceType == "MultipleChoice":
                result_answer = current_result.get(question, []) + [answer]
            elif choiceType == "Choice":
                result_answer = answer
            else:
                result_answer = self.get_concatenated_answer(
                    current_result, question, answer
                )
        elif choiceType == "Choice" and answer_else:
            result_answer = answer_else

        return result_answer

    @staticmethod
    def get_concatenated_answer(current_result, question, new_value):
        if current_value := current_result.get(question):
            return f"{current_value}, {new_value}"
        return new_value

    def get_agis_identify_response(self, markers, tolerance):
        base_url = build_url(
            settings.AG_GIS_BASE_URL,
            "/identify",
        )
        margin = Decimal(tolerance)
        x = Decimal(markers[0].get("x", ""))
        y = Decimal(markers[0].get("y", ""))
        payload = {
            "geometry": f"{x - margin},{y - margin},{x + margin},{y + margin}",
            "layers": "all",
            "tolerance": "0",
            "mapExtent": "2608986,1211613,2690520,1288912",
            "geometryType": "esriGeometryEnvelope",
            "imageDisplay": "10,10,96",
            "returnGeometry": "false",
            "f": "json",
        }
        response = self.session.post(
            base_url,
            data=payload,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

        try:
            response.raise_for_status()
        except requests.HTTPError as exc:  # pragma: no cover
            raise RuntimeError(
                _("An error occured while fetching data from the AGIS API.")
            ) from exc
        return response.json()

    def get_plot_data(self, agis_info):
        """Get data for "parzelle" table.

        Specific implementation to get list of plots
        containing information from _multiple_ layers,
        joined in a specific way.
        """
        result = []

        municipality = None
        municipality_info = self.get_layer_by_identifier(
            agis_info, "Gemeinden (Einzelfl√§chen)"
        )
        if municipality_info:
            municipality = municipality_info.get("value", None)

        plot_info = self.get_layer_by_identifier(agis_info, "AV: Parzellen")

        if plot_info:
            result.append(
                {
                    "gemeinde": municipality,
                    "parzellennummer": plot_info.get("attributes", {}).get(
                        "NUMMER", None
                    ),
                    "e-grid-nr": plot_info.get("attributes", {}).get("EGRID", None),
                }
            )
        return result

    def get_building_data(self, agis_info):
        """Get data for "gebaeude" table.

        Specific implementation to get list of buildings
        containing information from _multiple_ layers,
        joined in a specific way.
        """
        result = []

        building_info = self.get_layer_by_identifier(agis_info, "AV: Assekuranznummern")

        if building_info:
            result.append(
                {
                    "egid-nr": building_info.get("attributes", {}).get("EGID", None),
                    "amtliche-gebaeudenummer": building_info.get("attributes", {}).get(
                        "ASSNR", None
                    ),
                }
            )
        return result

    @staticmethod
    def get_layer_by_identifier(agis_info, identifier):
        return next(
            (r for r in agis_info if r["layerName"] == identifier),
            None,
        )

    @staticmethod
    def get_all_layers_by_identifier(agis_info, identifier):
        return [r for r in agis_info if r["layerName"] == identifier]

    @staticmethod
    def layer_exists(agis_info, identifiers):
        """Return true if one or more layers exists in agis_ingo from identifiers list, else false."""
        for layer in agis_info:
            if layer.get("layerName") in identifiers:
                return True
        return False
