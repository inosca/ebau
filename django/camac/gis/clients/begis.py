import concurrent.futures
import logging
from os import path
from urllib.parse import urlencode, urljoin

import requests
from django.conf import settings
from django.core.cache import cache
from lxml import etree

from camac.gis.clients.base import GISBaseClient

logger = logging.getLogger(__name__)


class BeGisClient(GISBaseClient):
    required_params = ["egrids"]
    is_queue_enabled = settings.BE_GIS_ENABLE_QUEUE

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.session: requests.Session = requests.Session()
        self.base_url = settings.GIS_BASE_URL

    def get_root(self, response):
        return etree.fromstring(response.content)

    def _build_polygon_url(self, egrid):
        path = "/geoservice3/services/a42geo/of_planningcadastre01_de_ms_wfs/MapServer/WFSServer"
        params = {
            "service": "WFS",
            "version": "2.0.0",
            "Request": "GetFeature",
            "typename": f"{settings.BE_GIS_POLYGON_SERVICE_CODE}:{settings.BE_GIS_POLYGON_LAYER_ID}",
            "count": 10,
            "Filter": f'<ogc:Filter><ogc:PropertyIsEqualTo matchCase="true"><ogc:PropertyName>EGRID</ogc:PropertyName><ogc:Literal>{egrid}</ogc:Literal></ogc:PropertyIsEqualTo></ogc:Filter>',
        }
        full_url = urljoin(self.base_url, path)
        query_string = urlencode(params)
        return f"{full_url}?{query_string}"

    def get_polygon(self, egrid):
        polygon_url = self._build_polygon_url(egrid)
        try:
            response = self.session.get(polygon_url)
            response.raise_for_status()
        except requests.exceptions.HTTPError as e:  # pragma: no cover
            logger.error(f"Polygon({egrid}): {e}")
            raise RuntimeError(
                f"Error {e.response.status_code} while fetching polygon data from the API"
            )
        except (
            requests.exceptions.Timeout,
            requests.exceptions.ConnectionError,
        ) as e:  # pragma: no cover
            logger.error(f"Polygon({egrid}): {e}")
            raise RuntimeError(
                "Connection error while fetching polygon data from the API"
            )

        try:
            root = self.get_root(response)
        except etree.XMLSyntaxError:  # pragma: no cover
            raise ValueError("Can't parse document")

        try:
            polygon = root.find(".//gml:Polygon", root.nsmap)
            polygon_to_string = etree.tostring(polygon, encoding="unicode")
            cache.set(egrid, polygon_to_string, 60)
            return polygon_to_string

        except (SyntaxError, TypeError) as e:
            raise ValueError("No polygon found") from e

    def get_url_data(self, egrid, service_code, boolean_layers, special_layers):
        polygon = cache.get(egrid) or self.get_polygon(
            egrid
        )  # checking if polygon already retrieved
        query = self.get_query(service_code, boolean_layers, special_layers, polygon)
        payload = self.get_feature_xml(service_code, query)
        try:
            response = self.session.post(
                f"{self.base_url}/geoservice3/services/a42geo/{service_code}/MapServer/WFSServer",
                data=payload,
            )
            response.raise_for_status()
            return response
        except requests.exceptions.HTTPError as e:  # pragma: no cover
            logger.error(f"{service_code}({egrid}): {e}")
            raise RuntimeError(
                f"Error {e.response.status_code} while fetching layer data from the API"
            ) from e
        except (
            requests.exceptions.Timeout,
            requests.exceptions.ConnectionError,
        ) as e:  # pragma: no cover
            logger.error(f"{service_code}({egrid}): {e}")
            raise RuntimeError(
                "Connection error while fetching layer data from the API"
            ) from e

    def send_requests_in_batches(
        self,
        data,
        egrids,
        batch_size,
        service_code,
        boolean_layers,
        special_layers,
        layers_dict,
    ):
        exception_messages = set()

        for i in range(0, len(egrids), batch_size):
            batch = egrids[i : i + batch_size]
            futures = []
            with concurrent.futures.ThreadPoolExecutor(
                max_workers=batch_size
            ) as executor:
                for egrid in batch:
                    futures.append(
                        executor.submit(
                            self.get_url_data,
                            egrid=egrid,
                            service_code=service_code,
                            boolean_layers=boolean_layers,
                            special_layers=special_layers,
                        )
                    )
                for future in concurrent.futures.as_completed(futures):
                    try:
                        response = future.result()
                        xml_data, et = self.get_xml_data(response)
                        new_data = self.get_data_from_xml(
                            service_code=service_code,
                            boolean_layers=boolean_layers,
                            special_layers=special_layers,
                            layers_dict=layers_dict,
                            xml_data=xml_data,
                            et=et,
                        )
                        self.merge_data_dict(data, new_data, special_layers)
                    except RuntimeError as e:  # pragma: no cover
                        exception_messages.add(str(e))

        if exception_messages:  # pragma: no cover
            # We raise a single RuntimeError per GIS datasource;
            # it includes unique error messages
            raise RuntimeError("/n".join(exception_messages))

        return data

    def process_data_source(self, config, _intermediate_data) -> dict:
        service_code = config.get("service_code")
        layers_dict = config.get("layers", {})
        boolean_layers, special_layers = self.get_config_layers(layers_dict)

        data = {}
        result = {}

        egrids = self.params.get("egrids").split(",")
        batch_size = settings.GIS_REQUESTS_BATCH_SIZE
        data = self.send_requests_in_batches(
            data=data,
            egrids=egrids,
            batch_size=batch_size,
            service_code=service_code,
            boolean_layers=boolean_layers,
            special_layers=special_layers,
            layers_dict=layers_dict,
        )

        for layer_id in special_layers:
            layer = layers_dict.get(layer_id, {})
            layer_data = data.get(layer_id)
            if layer_data:
                result.update(self.map_data(layer, layer_data, result))

        for layer_id in boolean_layers:
            layer = layers_dict.get(layer_id, {})
            layer_data = data.get(layer_id)
            result.update(self.map_data(layer, layer_data, result))

        return result

    def get_xml_data(self, response):
        try:
            et = self.get_root(response)
        except etree.XMLSyntaxError:  # pragma: no cover
            raise ValueError("Can't parse document")
        xml_data = et.findall("./gml:featureMember/", et.nsmap)
        return xml_data, et

    def get_data_from_xml(
        self, service_code, boolean_layers, special_layers, layers_dict, xml_data, et
    ):
        result = {}

        for layer_id in special_layers:
            layer = layers_dict.get(layer_id, {})
            search_term = layer.get("search_term", "")
            xml_layer_list = [child for child in xml_data if layer_id in child.tag]
            special_result_set = {
                item.text.strip()
                for child in xml_layer_list
                for item in child.findall(f"{service_code}:{search_term}", et.nsmap)
            }
            result[layer_id] = sorted(special_result_set)

        for layer_id in boolean_layers:
            xml_layer_list = [child for child in xml_data if layer_id in child.tag]
            result[layer_id] = len(xml_layer_list) > 0

        return result

    def merge_data_dict(self, data, new_data, special_layers):
        for layer in special_layers:
            new_layer_data = new_data.pop(layer)
            previous_layer_data = data.get(layer, [])
            data[layer] = sorted(set(new_layer_data + previous_layer_data))

        for key, value in new_data.items():
            if not data.get(key):
                data[key] = value
            elif data.get(key) and value:
                data[key] = value

    def map_data(self, layer, layer_data, intermediate_result):
        result = {}
        properties = layer["properties"]
        for property_config in properties:
            values = layer_data
            if property_config.get("mapper"):
                question = property_config["question"]
                values = getattr(self, f"map_{property_config['mapper']}")(
                    layer_data, intermediate_result, question
                )
                if values and not isinstance(values, list):
                    values = [values]

            if values:
                self.set_question_value(result, property_config["question"], values)

        return result

    def map_gewaesserschutzbereich_v2(self, values, intermediate_result, question):
        previous_value = intermediate_result.get(question, [])

        values_mapping = {
            "체briger Bereich 체B": "ueb",
            "Gew채sserschutzbereich Ao": "ao",
            "Gew채sserschutzbereich Au": "au",
        }

        values_to_add = [
            values_mapping[key] for key in values if key in values_mapping.keys()
        ]

        return (
            sorted(set(previous_value + values_to_add))
            if values_to_add
            else previous_value
        )

    def map_grundwasserschutzzonen_v2(self, values, intermediate_result, question):
        previous_value = intermediate_result.get(question, [])

        values_mapping = {
            "Grundwasserschutzzone S1": "s1",
            "Grundwasserschutzzone S2": "s2",
            "Grundwasserschutzzone S3": "s3-s3zu",
            "Grundwasserschutzzone S3Zu": "s3-s3zu",
            "Grundwasserschutzzone Sh": "sh",
            "Grundwasserschutzzone Sm": "sm",
            "Grundwasserschutzzone SA1": "sa",
            "Grundwasserschutzzone SA2": "sa",
            "Grundwasserschutzzone SA3": "sa",
            "Grundwasserschutzzone SBW": "sbw",
        }
        values_to_add = [
            values_mapping[key] for key in values if key in values_mapping.keys()
        ]

        return (
            sorted(set(previous_value + values_to_add))
            if values_to_add
            else previous_value
        )

    def map_boolean(self, values, intermediate_result, question):
        return "ja" if values else "nein"

    def map_nutzungszone(self, values, intermediate_result, question):
        previous_value = intermediate_result.get(question, [])

        return sorted(set(previous_value + values)) if values else previous_value

    def map_ueberbauungsordnung(self, values, intermediate_result, question):
        previous_value = intermediate_result.get(question, [])

        return sorted(set(previous_value + values)) if values else previous_value

    def get_config_layers(self, layers_dict):
        boolean_layers = [
            key
            for key, value in layers_dict.items()
            if value.get("is_boolean") and key not in settings.GIS_SKIP_BOOLEAN_LAYERS
        ]

        special_layers = [
            key
            for key, value in layers_dict.items()
            if not value.get("is_boolean")
            and key not in settings.GIS_SKIP_SPECIAL_LAYERS
        ]

        return boolean_layers, special_layers

    def get_query(self, service_code, boolean_layers, special_layers, polygon):
        template = """
            <Query typeName="{code}:{layer}" srsName="EPSG:2056">
                <ogc:Filter>
                    <ogc:Intersects>
                      {polygon}
                    </ogc:Intersects>
                </ogc:Filter>
            </Query>
        """
        query = "".join(
            map(
                lambda layer: template.format(
                    code=service_code, layer=layer, polygon=polygon
                ),
                boolean_layers + special_layers,
            )
        )
        return query

    def get_feature_xml(self, service_code, query):
        get_feature_xml = open(
            path.join(path.dirname(__file__), "xml/get_feature.xml"), "r"
        ).read()
        return get_feature_xml.format(
            baseURL=self.base_url, service_code=service_code, query=query
        )
