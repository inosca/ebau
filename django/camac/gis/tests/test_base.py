import pytest
from caluma.caluma_form.models import Question
from django.urls import reverse
from rest_framework import status

from camac.gis.clients.base import GISBaseClient
from camac.gis.models import GISDataSource


class FakeClient(GISBaseClient):
    def process_data_source(self, config, intermediate_data):
        return {
            "text-question": "foo",
            "table-question": [
                {
                    "table-question-1": "row 1 value 1",
                    "table-question-2": "row 1 value 2",
                },
                {
                    "table-question-1": "row 2 value 1",
                    "table-question-2": "row 2 value 2",
                },
            ],
        }


def test_process_data_source(db, gis_data_source):
    gis_client = GISBaseClient(GISDataSource.objects.all())
    fake_data = {}

    with pytest.raises(NotImplementedError):
        gis_client.process_data_source(gis_data_source, fake_data)


def test_view_structure(
    db,
    admin_client,
    gis_data_source_factory,
    mocker,
    question_factory,
    snapshot,
):
    question_factory(
        slug="text-question",
        label="Text Question",
        type=Question.TYPE_TEXT,
    )
    question_factory(
        slug="table-question",
        label="Table Question",
        type=Question.TYPE_TABLE,
        row_form__slug="table-form",
    )
    question_factory(
        slug="table-question-1",
        label="Question 1 in table",
        type=Question.TYPE_TEXT,
    )
    question_factory(
        slug="table-question-2",
        label="Question 2 in table",
        type=Question.TYPE_TEXT,
    )

    gis_data_source_factory()

    mocker.patch("camac.gis.views.get_client", return_value=FakeClient)
    mocker.patch("camac.gis.models.GISDataSource.get_required_params", return_value=[])

    response = admin_client.get(reverse("gis-data"))

    assert response.status_code == status.HTTP_200_OK
    snapshot.assert_match(response.json())
