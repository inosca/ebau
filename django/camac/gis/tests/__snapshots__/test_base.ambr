# serializer version: 1
# name: test_view_structure
  dict({
    'data': dict({
      'table-question': dict({
        'form': 'table-form',
        'hidden': False,
        'label': 'Table Question',
        'value': list([
          dict({
            'table-question-1': dict({
              'hidden': False,
              'label': 'Question 1 in table',
              'value': 'row 1 value 1',
            }),
            'table-question-2': dict({
              'hidden': False,
              'label': 'Question 2 in table',
              'value': 'row 1 value 2',
            }),
          }),
          dict({
            'table-question-1': dict({
              'hidden': False,
              'label': 'Question 1 in table',
              'value': 'row 2 value 1',
            }),
            'table-question-2': dict({
              'hidden': False,
              'label': 'Question 2 in table',
              'value': 'row 2 value 2',
            }),
          }),
        ]),
      }),
      'text-question': dict({
        'hidden': False,
        'label': 'Text Question',
        'value': 'foo',
      }),
    }),
  })
# ---
